### PR TITLE
[Relay] Regenerate socket URL on error disconnection

### PR DESCRIPTION
**Pull request by alexander-lsvk**
## Due Dilligence

* [ ] Breaking change
* [ ] Requires a documentation update

### CODE REVIEW

File: Sources/WalletConnectRelay/Dispatching.swift [hunk 1]
```diff
@@ -84,28 +102,51 @@ final class Dispatcher: NSObject, Dispatching {
...
+        setUpSocketConnectionObserving()
+    }
+    
+    private func setUpSocket() {
+        let socket = socketFactory.create(with: relayUrlFactory.create(
+            host: relayHost,
+            projectId: projectId
+        ))
+        socket.request.addValue(EnvironmentInfo.userAgent, forHTTPHeaderField: "User-Agent")
+        self.socket = socket
```
Comment (flypaper0) [inline Sources/WalletConnectRelay/Dispatching.swift:124]: why we need to recreate `socket` instance? Is it possible to just change `request` property?

File: Sources/WalletConnectRelay/Dispatching.swift [hunk 2]
```diff
@@ -84,28 +102,51 @@ final class Dispatcher: NSObject, Dispatching {
...
+        
+        switch socketConnectionType {
```
Comment (flypaper0) [inline Sources/WalletConnectRelay/Dispatching.swift:126]: Why need to recreate this on reconnection?

File: Sources/WalletConnectRelay/Dispatching.swift [hunk 3]
```diff
@@ -84,28 +102,51 @@ final class Dispatcher: NSObject, Dispatching {
...
 
     func disconnect(closeCode: URLSessionWebSocketTask.CloseCode) throws {
-        try socketConnectionHandler.handleDisconnect(closeCode: closeCode)
+        try socketConnectionHandler?.handleDisconnect(closeCode: closeCode)
+    }
+    
+    private func setUpSocketConnection() {
+        setUpSocket()
+        setUpWebSocketSession()
+        setUpSocketConnectionObserving()
```
Comment (flypaper0) [inline Sources/WalletConnectRelay/Dispatching.swift:115]: same question, is it necessary to `setUpSocketConnectionObserving` on reconnection?

File: Sources/WalletConnectRelay/Dispatching.swift [hunk 4]
```diff
@@ -84,28 +113,38 @@ final class Dispatcher: NSObject, Dispatching {
...
-        try socketConnectionHandler.handleConnect()
+        try socketConnectionHandler?.handleConnect()
     }
 
     func disconnect(closeCode: URLSessionWebSocketTask.CloseCode) throws {
-        try socketConnectionHandler.handleDisconnect(closeCode: closeCode)
+        try socketConnectionHandler?.handleDisconnect(closeCode: closeCode)
+    }
+    
+    private func setUpSocketConnection() {
```
Comment (llbartekll) [inline Sources/WalletConnectRelay/Dispatching.swift:123]: is this function called by anyone?

File: Sources/WalletConnectRelay/Dispatching.swift [hunk 5]
```diff
@@ -84,28 +113,38 @@ final class Dispatcher: NSObject, Dispatching {
...
             self.socketConnectionStatusPublisherSubject.send(.connected)
         }
-        socket.onDisconnect = { [unowned self] _ in
+        socket?.onDisconnect = { [unowned self] error in
             self.socketConnectionStatusPublisherSubject.send(.disconnected)
-
+            if error != nil {
+                self.socket?.request.url = relayUrlFactory.create(
+                    host: relayHost,
+                    projectId: projectId
```
Comment (llbartekll) [inline Sources/WalletConnectRelay/Dispatching.swift:143]: could we maybe add host and projectId to the url factory init? they wan't change anyway

File: Sources/WalletConnectRelay/RelayClient.swift [hunk 6]
```diff
@@ -80,24 +80,14 @@ public final class RelayClient {
         logger: ConsoleLogging = ConsoleLogger(loggingLevel: .off)
     ) {
         let clientIdStorage = ClientIdStorage(keychain: keychainStorage)
-        let socketAuthenticator = SocketAuthenticator(
+        let dispatcher = Dispatcher(
+            relayHost: relayHost,
+            projectId: projectId,
             clientIdStorage: clientIdStorage,
```
Comment (llbartekll) [inline Sources/WalletConnectRelay/RelayClient.swift:86]: would adding RelayClientFactory, like we have for other clients solve the init problem for Dispatcher and RelayClient?

File: Sources/WalletConnectRelay/Dispatching.swift [hunk 7]
```diff
@@ -13,10 +13,12 @@ protocol Dispatching {
 
 final class Dispatcher: NSObject, Dispatching {
     var onMessage: ((String) -> Void)?
-    var socket: WebSocketConnecting
-    var socketConnectionHandler: SocketConnectionHandler
-
+    var socket: WebSocketConnecting?
```
Comment (llbartekll) [inline Sources/WalletConnectRelay/Dispatching.swift:16]: does it have to be optional?

**Review by llbartekll — APPROVED** [review]
if tested manually and works then looks good to me

**Review by flypaper0 — APPROVED** [review]

**Comment by flypaper0** [comment]
lgtm, I think it's better to recreate websocket instance

**Pull request merged**